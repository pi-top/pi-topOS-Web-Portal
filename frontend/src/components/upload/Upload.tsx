import React from "react";
import DropZone from "react-dropzone";
import styles from "./Upload.module.css";

import { uploadFile } from "./utils/fileUpload";
import add from "../../assets/icons/add.svg";

export type Props = {
  onUploadStart?: () => void;
  onUploadEnd?: () => void;
  onUploadError?: (err: any) => void;
  onUploadProgress?: (progressEvent: ProgressEvent) => void;
  userInstruction: string;
  supportedExtensions: Array<string>;
  disabled: boolean;
  filenameRegex?: RegExp;
};

export default ({
  userInstruction,
  onUploadStart,
  onUploadEnd,
  onUploadError,
  onUploadProgress,
  supportedExtensions,
  filenameRegex,
  disabled,
}: Props) => {
  const handleFileUpload = async (file: File) => {
    try {
      if (file && filenameRegex && !filenameRegex.test(file.name)) {
        throw Error("Invalid filename");
      }
      await uploadFile(
        file,
        supportedExtensions,
        onUploadStart,
        onUploadEnd,
        onUploadProgress
      );
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      onUploadError && onUploadError(message);
      return;
    }
  };

  const onDrop = async (files: Array<File>) => {
    await handleFileUpload(files[0]);
  };

  return (
    <DropZone disabled={disabled} onDrop={onDrop}>
      {({ getRootProps, getInputProps }) => (
        <div
          data-testid="dropzone-input-wrapper"
          className={styles.dropzone}
          {...getRootProps()}
        >
          <input
            className={styles.dropzoneInput}
            data-testid="dropzone-input"
            disabled={disabled}
            {...getInputProps()}
          />
          <img src={add} alt="add" className={styles.addIcon} />
          <p data-testid="dropzone-instruction" className={styles.instruction}>{userInstruction}</p>
        </div>
      )}
    </DropZone>
  );
};
