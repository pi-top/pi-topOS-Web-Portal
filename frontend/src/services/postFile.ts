import axios from "axios";

import apiBaseUrl from "./apiBaseUrl";

export default async function postFile(
  data: FormData,
  onUploadProgress: (progressEvent: ProgressEvent) => void = (_) => {}
) {
  await axios.post(
    `${apiBaseUrl}/upload-file`,
    data,
    {
      onUploadProgress,
      headers: {
        "Content-Type": "multipart/form-data;",
      },
    }
  );
}
