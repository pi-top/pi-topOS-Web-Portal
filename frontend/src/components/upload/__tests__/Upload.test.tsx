import React from "react";
import {
  act,
  render,
  fireEvent,
  BoundFunction,
  QueryByBoundAttribute,
  wait,
} from "@testing-library/react";

import Upload from "../Upload";
import postFile from "../../../services/postFile";
jest.mock("../../../services/postFile");
const postFileMock = postFile as jest.Mock;

const createFile = (filename: string, size: number, type: string) => {
  const file = new File([""], filename, { type });
  Object.defineProperty(file, "size", {
    get() {
      return size;
    },
  });
  return file;
};

describe("Upload", () => {
  let defaultProps: Props;
  let container: HTMLElement;
  let queryByTestId: BoundFunction<QueryByBoundAttribute>;

  beforeEach(() => {
    postFileMock.mockResolvedValue("OK");

    defaultProps = {
        onUploadStart: jest.fn(),
        onUploadEnd: jest.fn(),
        onUploadError: jest.fn(),
        onUploadProgress: jest.fn(),
        userInstruction: "this is an instruction",
        supportedExtensions: ["zip"],
        disabled: false,
        filenameRegex: /^valid-filename\.zip$/,
    };

    ({ container, queryByTestId } =
      render(<Upload {...defaultProps} />));
  });

  it("renders provided instruction", () => {
    expect(queryByTestId("dropzone-instruction")).toMatchSnapshot();
  });

  it("renders a drop-zone to upload files", () => {
    expect(queryByTestId("dropzone-input-wrapper")).toMatchSnapshot();
  });

  it("calls onUploadError if user tries to upload a file with the wrong name or extension", async () => {
    act(() => {
      const dropzone = container.querySelector("input")!;
      fireEvent.change(dropzone, {
        target: {
          files: [createFile("not-a-zip.zip", 1234, "application/zip")],
        },
      });
    });

    await wait();
    await wait();
    expect(defaultProps.onUploadError).toHaveBeenCalledWith("Invalid filename");
  });

  it("calls onUploadError if user tries to upload a large file", async () => {
    act(() => {
      const dropzone = container.querySelector("input")!;
      fireEvent.change(dropzone, {
        target: {
          files: [
            createFile(
              "valid-filename.zip",
              2000000001,
              "application/zip"
            ),
          ],
        },
      });
    });

    await wait();
    await wait();
    expect(defaultProps.onUploadError).toHaveBeenCalledWith("Maximum upload file size exceeded");
  });

  it("calls onUploadError when upload process fails", async () => {
    postFileMock.mockRejectedValue(new Error("Oh oh"));

    act(() => {
      const dropzone = container.querySelector("input")!;
      fireEvent.change(dropzone, {
        target: {
          files: [
            createFile("valid-filename.zip", 100, "application/zip"),
          ],
        },
      });
    });

    await wait();
    await wait();
    expect(defaultProps.onUploadError).toHaveBeenCalled();
  });

  it("calls onUploadProgress while file is being uploaded", async () => {
    const progressData = { loaded: 10, total: 100 }
    postFileMock.mockImplementation(
      (
        data: FormData,
        onUploadProgress: (progressEvent: ProgressEvent) => void = (_) => {}
      ) => {
        return new Promise(() => {
          onUploadProgress(progressData);
        });
      }
    );
    act(() => {
      const dropzone = container.querySelector("input")!;
      fireEvent.change(dropzone, {
        target: {
          files: [
            createFile("valid-filename.zip", 100, "application/zip"),
          ],
        },
      });
    });

    await wait();
    await wait();
    expect(defaultProps.onUploadProgress).toHaveBeenCalledWith(progressData);

  });

  it("calls onUploadStart when a valid file is provided and upload starts", async () => {
    postFileMock.mockRejectedValue("Oh oh");

    act(() => {
      const dropzone = container.querySelector("input")!;
      fireEvent.change(dropzone, {
        target: {
          files: [
            createFile("valid-filename.zip", 100, "application/zip"),
          ],
        },
      });
    });

    await wait();
    await wait();
    expect(defaultProps.onUploadStart).toHaveBeenCalled();
  });

  it("calls onUploadEnd when upload completes", async () => {
    postFileMock.mockResolvedValue("OK");

    act(() => {
      const dropzone = container.querySelector("input")!;
      fireEvent.change(dropzone, {
        target: {
          files: [
            createFile("valid-filename.zip", 100, "application/zip"),
          ],
        },
      });
    });

    await wait();
    await wait();
    await wait();
    await wait();
    expect(defaultProps.onUploadEnd).toHaveBeenCalled();
  });
});
