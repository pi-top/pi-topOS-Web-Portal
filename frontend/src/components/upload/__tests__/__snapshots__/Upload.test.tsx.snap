// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Upload renders a drop-zone to upload files 1`] = `
<div
  class="dropzone"
  data-testid="dropzone-input-wrapper"
  role="presentation"
  tabindex="0"
>
  <input
    class="dropzoneInput"
    data-testid="dropzone-input"
    multiple=""
    style="display: none;"
    tabindex="-1"
    type="file"
  />
  <img
    alt="add"
    class="addIcon"
    src="add.svg"
  />
  <p
    class="instruction"
    data-testid="dropzone-instruction"
  >
    this is an instruction
  </p>
</div>
`;

exports[`Upload renders provided instruction 1`] = `
<p
  class="instruction"
  data-testid="dropzone-instruction"
>
  this is an instruction
</p>
`;
